File: frontend/src/api/admin/referral.ts
import { apiClient } from '../client'
import type {
  CustomAffiliate,
  CustomReferralAdminConfig,
  CustomReferralAdminOverview,
  CustomReferralCommission,
  CustomReferralSettlementBatch,
  CustomReferralWithdrawal,
  PaginatedResponse
} from '@/types'

async function getOverview(): Promise<CustomReferralAdminOverview> {
  const { data } = await apiClient.get<CustomReferralAdminOverview>('/admin/referral/overview')
  return data
}

async function getSettings(): Promise<CustomReferralAdminConfig> {
  const { data } = await apiClient.get<CustomReferralAdminConfig>('/admin/referral/settings')
  return data
}

async function updateSettings(payload: {
  provider: string
  cookie_ttl_days: number
  default_rate: number
  settle_freeze_days: number
  min_withdraw_amount: number
  withdraw_fee: number
}): Promise<CustomReferralAdminConfig> {
  const { data } = await apiClient.put<CustomReferralAdminConfig>('/admin/referral/settings', payload)
  return data
}

async function listAffiliates(params?: {
  page?: number
  page_size?: number
  status?: string
  keyword?: string
}): Promise<PaginatedResponse<CustomAffiliate>> {
  const { data } = await apiClient.get<PaginatedResponse<CustomAffiliate>>('/admin/referral/affiliates', { params })
  return data
}

async function approveAffiliate(userId: number, payload?: { rate_override?: number | null }): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/approve`, payload ?? {})
  return data
}

async function disableAffiliate(userId: number, payload?: { reason?: string }): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/disable`, payload ?? {})
  return data
}

async function rejectAffiliate(userId: number, payload?: { reason?: string }): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/reject`, payload ?? {})
  return data
}

async function restoreAffiliate(userId: number): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/restore`)
  return data
}

async function freezeSettlement(userId: number, payload?: { reason?: string }): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/settlement/freeze`, payload ?? {})
  return data
}

async function restoreSettlement(userId: number): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/settlement/restore`)
  return data
}

async function freezeWithdrawal(userId: number, payload?: { reason?: string }): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/withdrawal/freeze`, payload ?? {})
  return data
}

async function restoreWithdrawal(userId: number): Promise<CustomAffiliate> {
  const { data } = await apiClient.post<CustomAffiliate>(`/admin/referral/affiliates/${userId}/withdrawal/restore`)
  return data
}

async function listCommissions(params?: {
  page?: number
  page_size?: number
  status?: string
}): Promise<PaginatedResponse<CustomReferralCommission>> {
  const { data } = await apiClient.get<PaginatedResponse<CustomReferralCommission>>('/admin/referral/commissions', { params })
  return data
}

async function runSettlementBatch(): Promise<CustomReferralSettlementBatch> {
  const { data } = await apiClient.post<CustomReferralSettlementBatch>('/admin/referral/settlements/run')
  return data
}

async function listWithdrawals(params?: {
  page?: number
  page_size?: number
  status?: string
}): Promise<PaginatedResponse<CustomReferralWithdrawal>> {
  const { data } = await apiClient.get<PaginatedResponse<CustomReferralWithdrawal>>('/admin/referral/withdrawals', { params })
  return data
}

async function approveWithdrawal(id: number, payload?: { admin_note?: string }): Promise<CustomReferralWithdrawal> {
  const { data } = await apiClient.post<CustomReferralWithdrawal>(`/admin/referral/withdrawals/${id}/approve`, payload ?? {})
  return data
}

async function rejectWithdrawal(id: number, payload: { admin_note?: string; reject_reason: string }): Promise<CustomReferralWithdrawal> {
  const { data } = await apiClient.post<CustomReferralWithdrawal>(`/admin/referral/withdrawals/${id}/reject`, payload)
  return data
}

async function markWithdrawalPaid(id: number, payload?: {
  admin_note?: string
  payment_proof_url?: string
  payment_txn_no?: string
}): Promise<CustomReferralWithdrawal> {
  const { data } = await apiClient.post<CustomReferralWithdrawal>(`/admin/referral/withdrawals/${id}/pay`, payload ?? {})
  return data
}

async function uploadAsset(file: File): Promise<{ url: string }> {
  const formData = new FormData()
  formData.append('file', file)
  const { data } = await apiClient.post<{ url: string }>('/admin/referral/upload', formData)
  return data
}

export const adminReferralAPI = {
  getOverview,
  getSettings,
  updateSettings,
  listAffiliates,
  approveAffiliate,
  rejectAffiliate,
  disableAffiliate,
  restoreAffiliate,
  freezeSettlement,
  restoreSettlement,
  freezeWithdrawal,
  restoreWithdrawal,
  listCommissions,
  runSettlementBatch,
  listWithdrawals,
  approveWithdrawal,
  rejectWithdrawal,
  markWithdrawalPaid,
  uploadAsset
}

export default adminReferralAPI

File: backend/internal/server/routes/referral.go
package routes

import (
	"time"

	"github.com/Wei-Shaw/sub2api/internal/handler"
	ratelimit "github.com/Wei-Shaw/sub2api/internal/middleware"
	"github.com/Wei-Shaw/sub2api/internal/server/middleware"
	"github.com/Wei-Shaw/sub2api/internal/service"

	"github.com/gin-gonic/gin"
	"github.com/redis/go-redis/v9"
)

func RegisterReferralRoutes(
	r *gin.Engine,
	v1 *gin.RouterGroup,
	h *handler.Handlers,
	jwtAuth middleware.JWTAuthMiddleware,
	adminAuth middleware.AdminAuthMiddleware,
	settingService *service.SettingService,
	redisClient *redis.Client,
) {
	if h == nil || h.Referral == nil || h.Admin == nil || h.Admin.Referral == nil {
		return
	}

	rateLimiter := ratelimit.NewRateLimiter(redisClient)
	r.GET("/referral-assets/*path", h.Referral.ServeAsset)
	r.GET("/r/:code", rateLimiter.LimitWithOptions("custom-referral-landing", 60, time.Minute, ratelimit.RateLimitOptions{
		FailureMode: ratelimit.RateLimitFailClose,
	}), h.Referral.CaptureReferral)

	authenticated := v1.Group("")
	authenticated.Use(gin.HandlerFunc(jwtAuth))
	authenticated.Use(middleware.BackendModeUserGuard(settingService))
	{
		ext := authenticated.Group("/ext")
		{
			referral := ext.Group("/referral")
			{
				referral.GET("/profile", h.Referral.GetProfile)
				referral.GET("/summary", h.Referral.GetSummary)
				referral.POST("/apply", h.Referral.ApplyAffiliate)
				referral.GET("/commissions", h.Referral.ListCommissions)
				referral.GET("/withdrawals", h.Referral.ListWithdrawals)
				referral.POST("/withdrawals", h.Referral.CreateWithdrawal)
				referral.POST("/withdrawals/:id/cancel", h.Referral.CancelWithdrawal)
				referral.POST("/upload", h.Referral.UploadAsset)
			}
		}
	}

	admin := v1.Group("/admin")
	admin.Use(gin.HandlerFunc(adminAuth))
	{
		referral := admin.Group("/referral")
		{
			referral.GET("/overview", h.Admin.Referral.Overview)
			referral.GET("/settings", h.Admin.Referral.GetSettings)
			referral.PUT("/settings", h.Admin.Referral.UpdateSettings)
			referral.GET("/affiliates", h.Admin.Referral.ListAffiliates)
			referral.GET("/affiliates/:user_id/bindings", h.Admin.Referral.ListAffiliateBindings)
			referral.GET("/commissions", h.Admin.Referral.ListCommissions)
			referral.GET("/commission-jobs", h.Admin.Referral.ListCommissionJobs)
			referral.POST("/commissions/reverse", h.Admin.Referral.ReverseCommission)
			referral.POST("/affiliates/:user_id/approve", h.Admin.Referral.ApproveAffiliate)
			referral.POST("/affiliates/:user_id/rate", h.Admin.Referral.SetAffiliateRateOverride)
			referral.POST("/affiliates/:user_id/reject", h.Admin.Referral.RejectAffiliate)
			referral.POST("/affiliates/:user_id/disable", h.Admin.Referral.DisableAffiliate)
			referral.POST("/affiliates/:user_id/restore", h.Admin.Referral.RestoreAffiliate)
			referral.POST("/affiliates/:user_id/adjust", h.Admin.Referral.AdjustAffiliate)
			referral.POST("/affiliates/:user_id/settlement/freeze", h.Admin.Referral.FreezeSettlement)
			referral.POST("/affiliates/:user_id/settlement/restore", h.Admin.Referral.RestoreSettlement)
			referral.POST("/affiliates/:user_id/withdrawal/freeze", h.Admin.Referral.FreezeWithdrawal)
			referral.POST("/affiliates/:user_id/withdrawal/restore", h.Admin.Referral.RestoreWithdrawal)
			referral.POST("/settlements/run", h.Admin.Referral.RunSettlementBatch)
			referral.GET("/withdrawals", h.Admin.Referral.ListWithdrawals)
			referral.POST("/withdrawals/:id/approve", h.Admin.Referral.ApproveWithdrawal)
			referral.POST("/withdrawals/:id/reject", h.Admin.Referral.RejectWithdrawal)
			referral.POST("/withdrawals/:id/pay", h.Admin.Referral.MarkWithdrawalPaid)
			referral.POST("/upload", h.Admin.Referral.UploadAsset)
		}
	}
}
